arduino-ide-extension/src/browser/contributions/save-as-sketch.ts

```diff
@@ -84,91 +142,157 @@ export class SaveAsSketch extends SketchContribution {
     const defaultUri = containerDirUri.resolve(
-      exists
-        ? `${sketch.name}_copy_${dateFormat(new Date(), 'yyyymmddHHMMss')}`
-        : sketch.name
+      Sketch.toValidSketchFolderName(sketch.name, exists)
     );
     const defaultPath = await this.fileService.fsPath(defaultUri);
-    const { filePath, canceled } = await remote.dialog.showSaveDialog(
-      remote.getCurrentWindow(),
-      {
-        title: nls.localize(
-          'arduino/sketch/saveFolderAs',
-          'Save sketch folder as...'
-        ),
-        defaultPath,
+    return await this.promptLocalSketchFolderDestination(sketch, defaultPath);
+  }
+
+  /**
+   * Prompts for the new sketch folder name until a valid one is give,
+   * then resolves with the destination sketch folder URI string,
+   * or `undefined` if the operation was canceled.
+   */
+  private async promptLocalSketchFolderDestination(
+    sketch: Sketch,
+    defaultPath: string
+  ): Promise<string | undefined> {
+    let sketchFolderDestinationUri: string | undefined;
+    while (!sketchFolderDestinationUri) {
+      const { filePath } = await remote.dialog.showSaveDialog(
+        remote.getCurrentWindow(),
+        {
+          title: nls.localize(
+            'arduino/sketch/saveFolderAs',
+            'Save sketch folder as...'
+          ),
+          defaultPath,
+        }
+      );
+      if (!filePath) {
+        return undefined;
       }
-    );
-    if (!filePath || canceled) {
-      return false;
-    }
-    const destinationUri = await this.fileSystemExt.getUri(filePath);
-    if (!destinationUri) {
-      return false;
-    }
-    const workspaceUri = await this.sketchService.copy(sketch, {
-      destinationUri,
-    });
-    if (workspaceUri) {
-      await this.saveOntoCopiedSketch(sketch.mainFileUri, sketch.uri, workspaceUri);
-      if (markAsRecentlyOpened) {
-        this.sketchService.markAsRecentlyOpened(workspaceUri);
+      const destinationUri = await this.fileSystemExt.getUri(filePath);
+      // The new location of the sketch cannot be inside the location of current sketch.
+      // https://github.com/arduino/arduino-ide/issues/1882
+      let dialogContent: InvalidSketchFolderDialogContent | undefined;
+      if (new URI(sketch.uri).isEqualOrParent(new URI(destinationUri))) {
+        dialogContent = {
+          message: nls.localize(
+            'arduino/sketch/invalidSketchFolderLocationMessage',
+            "Invalid sketch folder location: '{0}'",
+            filePath
+          ),
+          details: nls.localize(
+            'arduino/sketch/invalidSketchFolderLocationDetails',
+            'You cannot save a sketch into a folder inside itself. This would go on forever.'
+          ),
+          question: nls.localize(
+            'arduino/sketch/editInvalidSketchFolderLocationQuestion',
+            'Do you want to try to save the sketch to a different location?'
```

Comment:
```suggestion
            'Do you want to try saving the sketch to a different location?'
```

The previous wording is fine, but I was slightly bothered by the three occurrences of the word "to". But only slightly bothered, so I'm OK if you want to reject this suggestion.

Reply:
Thank you!